README.md
<!--
 * @Author: nicheface nicheface@outlook.com
 * @Date: 2023-09-15 14:29:41
 * @LastEditors: nicheface nicheface@outlook.com
 * @LastEditTime: 2023-09-22 10:59:58
 * @FilePath: \\variable-print\\README.md
-->
<!--
 * @Author: gyg nicheface@outlook.com
 * @Date: 2023-09-15 14:29:41
 * @LastEditors: nicheface nicheface@outlook.com
 * @LastEditTime: 2023-09-22 09:49:13
 * @FilePath: \\variable-print\\README.md
-->
# variable-print

简体中文|[english](./README%20EN.md)

<div align=center>
	<img src="https://z1.ax1x.com/2023/09/22/pPoVJBR.png"/>
</div>

[![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/nicheface.variable-print?label=version)](https://marketplace.visualstudio.com/items?itemName=nicheface.variable-print)
[![Visual Studio Marketplace Downloads](https://img.shields.io/visual-studio-marketplace/d/nicheface.variable-print?label=downloads)](https://marketplace.visualstudio.com/items?itemName=nicheface.variable-print)
[![stars](https://img.shields.io/github/stars/nicheface/variable-print)](https://github.com/nicheface/variable-print.git)
[![issues](https://img.shields.io/github/issues/nicheface/variable-print)](https://github.com/nicheface/variable-print.git)
![GitHub](https://img.shields.io/github/license/nicheface/variable-print)

📕 [GitHub仓库](https://github.com/nicheface/variable-print.git)
📗 [VS Code插件市场](https://marketplace.visualstudio.com/items?itemName=nicheface.variable-print)
📘 [更新日志](https://github.com/nicheface/variable-print/blob/main/CHANGELOG.md)

📙 插件不断完善，欢迎提交 [Issues](https://github.com/nicheface/variable-print/issues)、[Pull requests](https://github.com/nicheface/variable-print/pulls)

---
>变量打印，自定义任意语言的变量打印，可以在扩展设置里面自定义修改
支持一行内多变量(逗号或者空格隔开)打印，支持多行多变量打印，**支持按住 <kbd>alt</kbd> 多选多个光标同时打印**

- [variable-print](#variable-print)
  - [使用说明](#使用说明)
    - [演示效果](#演示效果)
  - [配置](#配置)
    - [打印语句支持修改](#打印语句支持修改)
  - [更新日志](#更新日志)

## 使用说明

选中要打印的变量，按快捷键或者输入命令或者右击鼠标，选择`variable: 打印选中的变量`

- win,mac 快捷键 <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>p</kbd>
- <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>p</kbd>的命令是`Print Selected Variable`
- 鼠标右键，点击`variable: 打印选中的变量`

### 演示效果

<!-- ![recordprint](./imags/recoreprint.gif) -->

![recoreprint.gif](https://s1.locimg.com/2023/09/21/7003870674ca4.gif)

## 配置

>*用户配置优先级高于默认配置*

- `variable-print.userCustomPrintStatements` 是用户配置
- `variable-print.customPrintStatements` 是默认配置

### 打印语句支持修改

- `$v`代表当前变量(多个变量自动拆分)
- `$row`代表当前行
- `$col`代表当前列
- 支持多个后缀`h|c|cpp`使用同种配置

<!-- ![config](./imags/config.png) -->
[![pPokZ0x.png](https://z1.ax1x.com/2023/09/22/pPokZ0x.png)](https://imgse.com/i/pPokZ0x)

## 更新日志

- 2023-09-22 2.0.0 支持多光标，插入打印语句变成了异步操作，正确计算行号和列号，增加了用户配置选项，增加了vscode-nls，增加鼠标右键上下文，修改了license，完善了readme。
- 2023-09-16 1.0.3 增加了更多语言的默认配置并且终于完成了github actions的release.yml编写
- 2023-09-15 1.0.2 加了logo和github action
- 2023-09-15 1.0.1 改进了readme
- 2023-09-15 1.0.0 初始版本 可以自定义打印格式，在扩展设置里修改，支持 c|cpp|h 多个后缀名使用同种打印配置

**Enjoy!**


CHANGELOG.md
# Change Log 更新日志

All notable changes to the "variable-print" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [2.0.0] - 2023-09-22

### Added

- ✨ 支持多光标 (Multi-cursor support)
- ✨ 增加鼠标右键上下文 (Added right-click context menu)
- ✨ 增加了vscode-nls (Added vscode-nls)

### Changed

- 🎨 插入打印语句变成了异步操作 (Inserting print statements became asynchronous)
- 🎨 增加了用户配置选项 (Added user configuration options)
- 📖 修改了license (Updated license)

### Fixed

- 🐛正确计算行号和列号 (Correctly calculated line and column numbers)
- 📖完善了readme (Improved readme)

## [1.0.3] - 2023-09-16

### Added

- 新功能1：可以自动编译了发布了 (New feature 1: Can now compile and release automatically)

### Changed

- 变更1：加了更多语言的默认配置 (Change 1: Added more default configurations for different languages)

### Fixed

- 修复1：修复了某些语言默认配置 (Fixed issue 1: Fixed some default configurations for certain languages)

## [Unreleased]

- Initial release


src/extension.ts
/*
 * @Author: nicheface nicheface@outlook.com
 * @Date: 2023-09-15 14:29:41
 * @LastEditors: nicheface nicheface@outlook.com
 * @LastEditTime: 2023-09-21 14:46:21
 * @FilePath: \\variable-print\\src\\extension.ts
 */

import * as vscode from 'vscode';


function getCustomPrintStatement(languageId: string,id:number): string {
    const config = vscode.workspace.getConfiguration('variable-print');
    let key='';
    if (id===0){
        key='userCustomPrintStatements';
    }else if(id===1){
        key='customPrintStatements';
    }
    const customPrintStatements = config.get(key) as Record<string, string>;
    for (const key of Object.keys(customPrintStatements)) {
        const languages = key.split('|'); // 拆分用户或者默认定义的语言类型
        if (languages.includes(languageId)) {
            return customPrintStatements[key];
        }
    }
    return '';
}
function customSplit(input: string): string[] {
    const result: string[] = [];

    let currentLine = '';
    for (let i = 0; i < input.length; i++) {
        const char = input.charAt(i);

        if (char === '\n') {
            // 如果是换行符 "\n"，将当前行添加到结果数组中
            if (currentLine.length > 0) {
                result.push(currentLine);
            }
            result.push(char);
            currentLine = '';
        } else if (char === ',' || char === ' ' || char === '\t') {
            // 如果是逗号、空格或制表符，则将当前行添加到结果数组中
            if (currentLine.length > 0) {
                result.push(currentLine);
            }
            currentLine = '';
        } else if (char !== '\r') {
            // 如果不是回车符 "\r"，将字符添加到当前行
            currentLine += char;
        }
    }

    // 如果还有剩余的内容，将其添加到结果数组中
    if (currentLine.length > 0) {
        result.push(currentLine);
    }

    return result;
}

function insertCustomPrintStatement(editor: vscode.TextEditor) {
    const { selections } = editor;
    const activeEditor = vscode.window.activeTextEditor;
    if (!activeEditor) {
        vscode.window.showErrorMessage('no active editor');
        return;
    }

    const { languageId } = editor.document;
    const userCustomPrintStatement = getCustomPrintStatement(languageId,0);
    let printvariableconfig=userCustomPrintStatement;
    if (!userCustomPrintStatement) {
        const customPrintStatement = getCustomPrintStatement(languageId,1);
        printvariableconfig=customPrintStatement;
        if (!customPrintStatement) {
            vscode.window.showErrorMessage(`No custom print statement found for language ${languageId}`);
            return;
        }
    }
    

    let total = 0;
    let temprow = 0;

    const workspaceEdit = new vscode.WorkspaceEdit();

    const selectionsArray = [...selections]; // 将 Selection 数组转换为普通数组
    const sortedSelections = selectionsArray.sort((a, b) => {
        return a.start.line - b.start.line;
    });
    for (let [i, selection] of sortedSelections.entries()) {
        const startLine = selection.start.line;
        const startCharacter = selection.start.character;
        const endLine = selection.end.line;
        const currentIndentation = editor.document.lineAt(startLine).text.match(/^\s*/)?.[0] || '';

        const selectedVariables = customSplit(editor.document.getText(selection));

        let insertLines: string = '';
        let pre = 0;
        let currentrow = startLine;
        let currentcol = startCharacter;

        let singleselectedrow = 0;
        if (selectedVariables.length === 0) {
            vscode.window.showErrorMessage(`found blank selection at line ${currentrow + total + 1}`);
            continue;
        }
        for (const selectedVariable of selectedVariables) {
            if (selectedVariable === '\n') {
                currentrow += 1;
                currentcol = currentIndentation.length;
                pre = 0;
                continue;
            }
            currentcol += pre;
            pre += selectedVariable.length;
            const printStatement = `${currentIndentation}${printvariableconfig
                .replace(/\$v/g, selectedVariable)
                .replace(/\$row/g, (currentrow + total + 1).toString())
                .replace(/\$col/g, (currentcol + 1).toString())}\n`;

            insertLines += printStatement;
            singleselectedrow += 1;

        }
        if (i !== sortedSelections.length - 1) {
            if (sortedSelections[i + 1].end.line === sortedSelections[i].end.line) {
                temprow += singleselectedrow;
            } else {
                total = total + temprow + singleselectedrow;
                temprow = 0;
            }
        }
        // 将编辑操作添加到 workspaceEdit 中
        const newPosition = new vscode.Position(endLine + 1, 0);
        if (endLine === editor.document.lineCount - 1) {
            workspaceEdit.insert(editor.document.uri, new vscode.Position(endLine + 1, 0), '\r');
        }
        workspaceEdit.insert(editor.document.uri, newPosition, insertLines);
    }
    vscode.workspace.applyEdit(workspaceEdit);
}

export function activate(context: vscode.ExtensionContext) {
    

    const userConfig = vscode.workspace.getConfiguration('variable-print');
    let userCustomPrintStatements = userConfig.get('userCustomPrintStatements') || {};
    if(!userCustomPrintStatements){
        // 更新用户配置
        // eslint-disable-next-line @typescript-eslint/naming-convention
        userConfig.update('userCustomPrintStatements', {"":""}, vscode.ConfigurationTarget.Global);
        vscode.window.showInformationMessage('user configuration has been reset.');
    }
    const disposable = vscode.commands.registerCommand('extension.printVariable', () => {
        // vscode.window.showInformationMessage(printtitle);
        const editor = vscode.window.activeTextEditor;
        if (editor) {
            const selectedText = editor.document.getText(editor.selection);
            if (selectedText) {
                insertCustomPrintStatement(editor);
            }
        }
    });

    context.subscriptions.push(disposable);
}

export function deactivate() { }


